storage/DataProviderThread.hpp

```diff
@@ -0,0 +1,103 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ **/
+
+#ifndef QUICKSTEP_STORAGE_DATA_PROVIDER_THREAD_HPP_
+#define QUICKSTEP_STORAGE_DATA_PROVIDER_THREAD_HPP_
+
+#include <cstddef>
+
+#include "query_execution/QueryExecutionTypedefs.hpp"
+#include "storage/StorageConfig.h"
+#include "threading/Thread.hpp"
+#include "utility/Macros.hpp"
+
+#include "tmb/native_net_client_message_bus.h"
+
+namespace quickstep {
+
+class CatalogDatabase;
+class CatalogRelation;
+class QueryProcessor;
+class StorageManager;
+
+/** \addtogroup Storage
+ *  @{
+ */
+
+/**
+ * @brief A thread that provides access to query results (e.g. storage blocks),
+ *        to Quickstep clients.
```

Comment:
Could you please provide more info regarding this class, including which class will own this class, and why it needs write permissions to both `CatalogDatabase` and `QueryProcessor`? Thanks!

Reply:
The DataProviderThread class will be part of the Quickstep server process. The following image describes the role of this thread. 
![image](https://user-images.githubusercontent.com/3750801/40525388-5e6225da-5fa5-11e8-818c-fc6f913f471c.png)

Reply:
This thread will be responsible to drop temporary tables created by queries. To perform this drop operation, it needs access to both CatalogDatabase and QueryProcessor. Details in the code [here](https://github.com/hbdeshmukh/incubator-quickstep/blob/6967ca57c448a17b63e176bf8bc1576066fe893a/storage/DataProviderThread.cpp#L79)